test/json/validation/points.markets-listed.test.ts

```diff
@@ -0,0 +1,135 @@
+ * so we can assert their presence locally without a network call.
+ */
+const shouldSkipApiTests = process.env.SKIP_MARKETS_API_TESTS === "true";
+
+// Morpho GraphQL endpoint
+const MORPHO_API_URL = "https://api.morpho.org/graphql";
+
+// Helper to query marketByUniqueKey
+async function fetchMarketByUniqueKey(
+  uniqueKey: string,
+  chainId: number,
+): Promise<{ uniqueKey: string; whitelisted: boolean } | null> {
+  const query = `
+    query MarketByUniqueKey($uniqueKey: String!, $chainId: Int!) {
+      marketByUniqueKey(uniqueKey: $uniqueKey, chainId: $chainId) {
+        uniqueKey
+        whitelisted
+      }
+    }
+  `;
+
+  const res = await fetch(MORPHO_API_URL, {
+    method: "POST",
+    headers: { "content-type": "application/json" },
+    body: JSON.stringify({
+      query,
+      variables: { uniqueKey, chainId },
+    }),
+  });
+
+  if (!res.ok) {
+    throw new Error(
+      `Morpho API HTTP error ${res.status} for market ${uniqueKey} on chain ${chainId}`,
+    );
+  }
+
+  const json = await res.json();
+
+  if (json.errors?.length) {
+    const msg = json.errors.map((e: any) => e.message).join("; ");
+    throw new Error(
+      `Morpho API GraphQL error for market ${uniqueKey} on chain ${chainId}: ${msg}`,
+    );
+  }
+
+  return json.data?.marketByUniqueKey ?? null;
+}
+
+// Collect all market unique keys from both relevant maps
+function collectAllMarketIds(points: PointsMapping): Record<string, Set<string>> {
+  const byChain: Record<string, Set<string>> = {};
+
+  const add = (chainIdStr: string, marketId: string) => {
+    if (!byChain[chainIdStr]) byChain[chainIdStr] = new Set<string>();
+    byChain[chainIdStr].add(marketId);
+  };
+
+  // 1) marketsWithPoints: keys are market unique keys
+  for (const [chainIdStr, marketsMap] of Object.entries(points.marketsWithPoints)) {
+    for (const marketId of Object.keys(marketsMap)) {
+      add(chainIdStr, marketId);
+    }
+  }
+
+  // 2) vaultsWithPointsOnMarket: keys are also market unique keys
+  for (const [chainIdStr, marketsMap] of Object.entries(points.vaultsWithPointsOnMarket)) {
+    for (const marketId of Object.keys(marketsMap)) {
+      add(chainIdStr, marketId);
+    }
+  }
+
+  return byChain;
+}
+
+const allMarketIdsByChain = collectAllMarketIds(points);
+
+describe("points.json – market unique keys correspond to real Morpho markets", () => {
+  if (shouldSkipApiTests) {
+    it.skip("SKIPPED via SKIP_MARKETS_API_TESTS", () => {});
+    return;
+  }
+
+  // Allow more time for network calls
+  jest.setTimeout(60_000);
+
+  for (const [chainIdStr, marketIds] of Object.entries(allMarketIdsByChain)) {
+    const chainId = Number(chainIdStr);
+
+    for (const marketUniqueKey of marketIds) {
+      it(
+        `market ${marketUniqueKey} on chain ${chainId} exists (Morpho API)`,
+        async () => {
+          const market = await fetchMarketByUniqueKey(marketUniqueKey, chainId);
```

Comment:
This loop will make hundreds of requests to the API (one per marketId). One nice improvement would be to use the `markets` query to get them all in one go (or e.g. one request per chain) and compare against that list.

Regardless, this is awesome work!